lib/std/debug.zig

```diff
@@ -50,163 +40,378 @@ pub const LineInfo = struct {
+                return .escape_codes;
+            } else if (builtin.os.tag == .windows and stderr_file.isTty()) {
+                return .windows_api;
+            } else {
+                return .no_color;
+            }
+        }
+    }
+
+    pub const AddressSymbolMapping = struct {
+        debug_info: DebugInfo,
+
+        pub fn addressToSymbol(self: *@This(), address: usize) !SymbolInfo {
+            const module = self.debug_info.getModuleForAddress(address) catch |err| switch (err) {
+                error.MissingDebugInfo, error.InvalidDebugInfo => {
+                    return SymbolInfo{};
+                },
+                else => return err,
+            };
+
+            return module.addressToSymbol(address);
+        }
+
+        pub fn init(allocator: *std.mem.Allocator) !@This() {
+            return @This(){ .debug_info = try openSelfDebugInfo(allocator) };
+        }
+    };
+
+    fn writeLineFromFileAnyOs(writer: anytype, line_info: LineInfo) !void {
+        // Need this to always block even in async I/O mode, because this could potentially
+        // be called from e.g. the event loop code crashing.
+        var f = try fs.cwd().openFile(line_info.file_name, .{ .intended_io_mode = .blocking });
+        defer f.close();
+        // TODO fstat and make sure that the file has the correct size
+
+        var buf: [mem.page_size]u8 = undefined;
+        var line: usize = 1;
+        var column: usize = 1;
+        var abs_index: usize = 0;
+        while (true) {
+            const amt_read = try f.read(buf[0..]);
+            const slice = buf[0..amt_read];
+
+            for (slice) |byte| {
+                if (line == line_info.line) {
+                    try writer.writeByte(byte);
+                    if (byte == '\n') {
+                        return;
+                    }
+                }
+                if (byte == '\n') {
+                    line += 1;
+                    column = 1;
+                } else {
+                    column += 1;
+                }
+            }
+
+            if (amt_read < buf.len) return error.EndOfFile;
+        }
+    }
+
+    pub fn attemptWriteLineFromSourceFile(writer: anytype, line_info: LineInfo) !bool {
+        writeLineFromFileAnyOs(writer, line_info) catch |err| {
+            switch (err) {
+                error.EndOfFile, error.FileNotFound => {},
+                error.BadPathName => {},
+                error.AccessDenied => {},
+                else => return err,
+            }
+            return false;
+        };
+        return true;
+    }
+
+    /// When overriding panicTerminate, if you *must* lock a mutex, be careful not to panic!
+    /// Otherwise, this will would call panicTerminate recursively and deadlock.
+    pub const panicTerminate = os.abort;
+};
+
+const is_default_config = @hasDecl(root, "debug_config");
+const config = if (is_default_config)
+    root.debug_config
+else
+    default_config;
```

Comment:
This PR uses the `root.debug_config` namespace to override functions. Given the large number of overrides, I think this is reasonable. At the moment, std.log *doesn't* use this approach, functions are defined totally adhoc (`logWriter`, `logDetectTTYConfig`, and the already existing `log` and `log_level`).

I think it might make sense to instead have a `root.log_config` namespace. Design questions:
 - Should this namespace exists instead of the existing `log` and `log_level` overrides (those values would be possible to  override from the config)?
 - Should they both exist? This is less breaking, but a bit gross and requires some checking.

Reply:
I like the `debug_config` idea, cluttering the root namespace with a thousand different functions is bad and may raise some unexpected problems (#6514).